scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java

```diff
@@ -431,7 +433,7 @@ public final void processMessage(Record record) throws HandshakeException {
 		// The DTLS 1.2 spec (section 4.1.2.6) advises to do replay detection
 		// before MAC validation based on the record's sequence numbers
 		// see http://tools.ietf.org/html/rfc6347#section-4.1.2.6
-		if (!session.isDuplicate(record.getSequenceNumber())) {
+		if (session.getReadEpoch() != record.getEpoch() || !session.isDuplicate(record.getSequenceNumber())) {
```

Comment:
:-) test it without :-). 

The first CLIENT_HELLO (epoch 0) will have the same seqn as the FINISH (epoch 1). 
So the check for duplicate (not taking the epoch in account) will filter out the FINISH.
With the VERIFY_REQUEST, the first CLIENT_HELLO is not processed here and therefore
"markRecordAsRead" is not called.

The other caller of `isDuplicate` is `isRecordProcessable`, which checks the epoch ahead.

And, sure, the "sliding window" implementation may be renewed. Though RFC6347 just mentions to update the window if the MAC is validated, it seems to be only defined for epoch 1 (or higher). And the check `sequenceNo > receiveWindowUpperBoundary` in `isDuplicate` may also be discussed. But this may end up in longer discussions about the dtls deduplication.
So I think, it's OK, to change it here. 

Just to mention: the implementation before this PR uses the same window for two epochs, which is also dangerous. If the FINISH must be resend, it may reach the seqn, which is already masked in the receive window for epoch 0.

Reply:
```
boolean sameEpoch = session.getReadEpoch() == record.getEpoch();
		if ((sameEpoch && !session.isDuplicate(record.getSequenceNumber()))
				|| (session.getReadEpoch() + 1) == record.getEpoch()) {
```

That new check should be closest to the old implementation but will not fail with the same seqn and different epochs.

Reply:
I agree, ideadly "sliding windows"  should taking into account epoch. Knowing that we do not allow renegociation, we have only  2 epochs (0 or 1). So maybe, just duplicate attributes about sliding windows (one for epoch 0 and one for epoch 1), could do the job.

Anyway, this check is better than before !

Reply:
https://tools.ietf.org/html/rfc6347#section-4.1.2.6

> The receive window is updated only if the MAC verification succeeds.

In this method, only handshake messages are processed. As long as epoch 0 is valid , epoch 1 records could not be verified. Therefore `markRecordAsRead` will not be effective for the epoch 1 handshake message, if it is queued and processed later.
After CCS with the finish handshake message, there is no more handshake message of epoch 1, which will comes afterwards. The only thing, which may be improved, is to pass that epoch 1 record to `markRecordAsRead` after that record is processed successful.